src/libs/safe-path/src/pinned_path_buf.rs

```diff
@@ -0,0 +1,384 @@
+    /// - fcntl(F_GETFD, F_SETFD, F_GETFL)
+    pub fn path_fd(&self) -> RawFd {
+        self.handle.as_raw_fd()
+    }
+
+    /// Get [Metadata] about the path handle.
+    pub fn metadata(&self) -> Result<Metadata> {
+        self.handle.metadata()
+    }
+
+    /// Create or open a child directory if current object is a directory.
+    pub fn mkdir(&self, path_comp: &OsStr, mode: libc::mode_t) -> Result<Self> {
+        let path_name = Self::prepare_path_component(path_comp)?;
+        let res = unsafe { libc::mkdirat(self.handle.as_raw_fd(), path_name.as_ptr(), mode) };
+        if res < 0 {
+            let e = Error::last_os_error();
+            if e.kind() != ErrorKind::AlreadyExists {
+                return Err(e);
+            }
+        }
+
+        self.open_child(path_comp)
+    }
+
+    /// Get the link target.
+    pub(crate) fn target(&self) -> &Path {
+        &self.target
+    }
+
+    /// Open a directory/file by path.
+    ///
+    /// Obtain a file descriptor that can be used for two purposes:
+    /// - indicate a location in the filesystem tree
+    /// - perform operations that act purely at the file descriptor level
+    fn open_by_path<P: AsRef<Path>>(path: P) -> Result<File> {
+        // When O_PATH is specified in flags, flag bits other than O_CLOEXEC, O_DIRECTORY, and
+        // O_NOFOLLOW are ignored.
+        let o_flags = libc::O_PATH | libc::O_CLOEXEC;
+        OpenOptions::new()
+            .read(true)
+            .custom_flags(o_flags)
+            .open(path.as_ref())
+    }
+
+    fn get_proc_path<F: AsRawFd>(file: F) -> PathBuf {
+        PathBuf::from(format!("/proc/self/fd/{}", file.as_raw_fd()))
+    }
+
+    fn new_from_file<P: AsRef<Path>>(handle: File, orig_path: P) -> Result<Self> {
+        let path = Self::get_proc_path(handle.as_raw_fd());
+        let link_path = fs::read_link(path.as_path())?;
+        if link_path != orig_path.as_ref() {
+            Err(Error::new(
+                ErrorKind::Other,
+                format!(
+                    "Path changed from {} to {} on open, possible attack",
+                    orig_path.as_ref().display(),
+                    link_path.display()
+                ),
+            ))
+        } else {
+            Ok(PinnedPathBuf {
+                handle,
+                path,
+                target: link_path,
+            })
+        }
+    }
+
+    #[inline]
+    fn prepare_path_component(path_comp: &OsStr) -> Result<CString> {
+        let path = Path::new(path_comp);
+        let mut comps = path.components();
+        let name = comps.next();
+        if !matches!(name, Some(Component::Normal(_))) || comps.next().is_some() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Path component {} is invalid", path_comp.to_string_lossy()),
+            ));
+        }
+        let name = name.unwrap();
+        if name.as_os_str() != path_comp {
```

Comment:
It's used to detect invalid parameters because Path::components() has special treatments for
```
When parsing the path, there is a small amount of normalization:
- Repeated separators are ignored, so a/b and a//b both have a and b as components.
- Occurrences of . are normalized away, except if they are at the beginning of the path. For example, a/./b, a/b/, a/b/. and a/b all have a and b as components, but ./a/b starts with an additional CurDir component.
- A trailing slash is normalized away, /a/b and /a/b/ are equivalent.
```

Reply:
Ok, I see.